### PR TITLE
feature: Plugin UI extensible area for user list dropdown and associated refactors

### PR DESCRIPTION
**Pull request by GuiLeme**
### What does this PR do?

It adds support  for the new extensible Area of `UserListDropdownItem` and also for the `useLoadedUserList` hook from the Plugin Architecture. See demo video below:

https://github.com/bigbluebutton/bigbluebutton/assets/69865537/ee4b5682-7a57-446a-aa50-2438db74c473

Lastly, it corrects some issues spotted from the previous iteration on the plugin architecture on the extensible area (PresentationToolbarPlugin).

### More

To run this PR follow the next steps:

**Clone the SDK repository to have a sample plugin to test this on**
``` bash
git clone https://github.com/bigbluebutton/bigbluebutton-html-plugin-sdk.git
```

**Download dependencies and start the plugin on development mode (Outside of the docker)**
``` bash
cd samples/sample-user-list-dropdown-plugin
npm install
npm start
```
Reminder: We are here testing the `user-list-dropdown-menu-plugin`, but feel free to test with the other sample.

**Adjust the settings.yml to call sample-user-list-dropdown-plugin**

``` bash
  plugins:
    - name: SampleUserListDropdownPlugin
      url: http://127.0.0.1:8080/static/SampleUserListDropdownPlugin.js
```

**Restart bigbluebutton-html5 from source**
 
These steps already serve this PR, but if there is an update on the SDK in the future, make sure to run instructions provided by @TiagoJacobs on https://github.com/bigbluebutton/bigbluebutton-html-plugin-sdk/pull/9#issuecomment-1697949116 (Those are needed if both client and SDK are being developed in parallel).

### CODE REVIEW

**Review by TiagoJacobs — CHANGES_REQUESTED** [review]

File: bigbluebutton-html5/imports/ui/components/user-list/user-list-content/user-participants/user-list-participants/user-actions/component.tsx [hunk 1]
```diff
@@ -31,7 +33,19 @@ interface UserActionsProps {
...
+
+interface DropdownItem {
+  key: string;
+  label: string | undefined;
+  icon: string | undefined;
+  tooltip: string | undefined;
+  allowed: boolean | undefined;
+  iconRight: string | undefined;
+  divider: boolean | undefined;
+  isOptionEmpty: boolean | undefined;
```
[the remark this answers is not in the document]
Comment (GuiLeme) [reply bigbluebutton-html5/imports/ui/components/user-list/user-list-content/user-participants/user-list-participants/user-actions/component.tsx:46]: I thought about it too, but it seems that letting `isOprionEmpty` gives us a more generic property, that could be used not only to mount a separator. But on the other hand, for now, we only have the separator option, so, what do you say?

File: bigbluebutton-html5/imports/ui/components/plugins-engine/plugin-provided-state/container.tsx [hunk 2]
```diff
@@ -35,24 +40,36 @@ const PluginProvidedStateContainer = (props: PluginProvidedStateContainerProps)
...
+    );
     setPluginsProvidedAggregatedState(
       {
         ...pluginsProvidedAggregatedState,
         presentationToolbarItems: aggregatedPresentationToolbarItems,
+        userListDropdownItems: aggregatedUserListDropdownItems,
       },
     );
-  }, [presentationToolbarItems]);
+  }, [presentationToolbarItems, userListDropdownItems]);
```
[the remark this answers is not in the document]
Comment (GuiLeme) [reply bigbluebutton-html5/imports/ui/components/plugins-engine/plugin-provided-state/container.tsx:62]: Hum, I don't know about that. I think the way it is written, it seems readable already and besides it avoids code duplicate (it wouldn't be much really, just the  `setPluginsProvidedAggregatedState` supposedly).

What we could do is separate those into two different files, like perhaps `plugin-provided-state/presentation-toolbar/container.tsx` and `plugin-provided-state/user-list-dropdown/container.tsx`. What do you say?

File: bigbluebutton-html5/imports/ui/components/user-list/user-list-content/user-participants/user-list-participants/user-actions/component.tsx (hunk 1 above)
[the remark this answers is not in the document]
Comment (TiagoJacobs) [reply bigbluebutton-html5/imports/ui/components/user-list/user-list-content/user-participants/user-list-participants/user-actions/component.tsx:46]: We have one property named `divider`, and another named `dividerTop`.

However the plugin is calling it as `separator`, and treating it as a different item, which I appreciate.

So I would propose that you modify the existing code to prepend an item like `{separator: true}` for the cases that are now using `dividerTop: true`. Same thing for appending when it's `divider: true`.

This way we clean-up these different terms in the existing code, and make it more compatible with how the plugin is working.

**Review by Tainan404 — CHANGES_REQUESTED** [review]

**Review by Tainan404 — APPROVED** [review]
LGTM

**Review by TiagoJacobs — APPROVED** [review]
LGTM

**Comment by TiagoJacobs** [comment]
Hello @GuiLeme - just for helping others in the future, can you remove this part of your PR description, and replace it by instructions on how to try it?

image

@ramonlsouza - after that you can try to run it, and if everything goes well, we can merge!

**Review by ramonlsouza — APPROVED** [review]

**Pull request merged**